portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.Login; 
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    // returns empty string if user is logged in and the login url if user is not logged in  
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        String loginUrl=null, logoutUrl=null;
+        String userEmail=null;
+        Boolean loggedIn=userService.isUserLoggedIn();
+
+        // user not logged in, return login link
+        if (!loggedIn) {
+            loginUrl = userService.createLoginURL("/");
+        }
+
+        // user logged in, return logout link
+        else {
+            logoutUrl = userService.createLogoutURL("/");
+            userEmail = userService.getCurrentUser().getEmail();
+        }
+
+        Login loginDetails = new Login(loggedIn, loginUrl, logoutUrl, userEmail);
+
+        response.setContentType("application/json");
+        response.getWriter().println(convertToJsonUsingGson(loginDetails));
+        return;
+    }
+
+    // convert to JSON using GSON
+    private String convertToJsonUsingGson(Login loginDetails) {
```

Comment:
This function name convertToJsonUsingGson is the same as the function in DataServlet.Java, but does slightly different thing, it might be better to move both functions to an Utility class that does Json conversion, so it is easy to understand that why you have two verify similar functions in the codebase.

Reply:
You can use generics in Java to create a function that can convert multiple classes to Json string. For example, take a look at my utility class.

https://github.com/kai-dream/kai-sps2020-portfolio/blob/kai-development/portfolio/src/main/java/com/google/sps/servlets/JsonUtility.java

Reply:
ok, moved to another utility class. thanks for the suggestion!